sql/src/main/java/io/crate/planner/operators/HashJoinConditionSymbolsExtractor.java

```diff
@@ -0,0 +1,116 @@
+ * algorithm. The extractor will detect if a symbol occurs in multiple EQ condition and if so won't extract any symbols
+ * from this EQ condition (skipping all EQ operator arguments).
+ * <p>
+ * Example:
+ * <pre>
+ *     t1.a = t2.a AND function(t1.a) = t2.b AND function(t1.a) = t2.c AND t1.x > t2.y
+ * </pre>
+ *
+ * will result in:
+ *
+ * <pre>
+ *  t1: [t1.a, function(t1.a)]
+ *  t2: [t2.a, t2.b]
+ *  </pre>
+ * <p>
+ * as t2.c is compared to the same symbol as t2.b and only EQ operators are processed.
+ */
+public class HashJoinConditionSymbolsExtractor {
+
+    private static final Visitor VISITOR = new Visitor();
+    private static final RelationDetector RELATION_DETECTOR = new RelationDetector();
+
+    /**
+     * Extracts all symbols per relations from any EQ join conditions. See {@link HashJoinConditionSymbolsExtractor}
+     * class documentation for details.
+     */
+    public static Map<AnalyzedRelation, List<Symbol>> extract(Symbol symbol) {
+        Map<AnalyzedRelation, List<Symbol>> symbolsPerRelation = new HashMap<>();
+        VISITOR.process(symbol, symbolsPerRelation);
+
+        return symbolsPerRelation;
+    }
+
+    public static class Visitor extends DefaultTraversalSymbolVisitor<Map<AnalyzedRelation, List<Symbol>>, Void> {
+
+        @Override
+        public Void visitFunction(Function function, Map<AnalyzedRelation, List<Symbol>> symbolsPerRelation) {
+            String functionName = function.info().ident().name();
+            if (functionName.equals(EqOperator.NAME)) {
+                int duplicatePos = 0;
+                for (Symbol arg : function.arguments()) {
+                    AnalyzedRelation relation = RELATION_DETECTOR.process(arg, null);
+                    List<Symbol> symbols = symbolsPerRelation.computeIfAbsent(relation, k -> new ArrayList<>());
+                    if (symbols.contains(arg)) {
+                        duplicatePos = symbols.size();
+                        continue;
+                    }
+                    symbols.add(arg);
+                }
+                // if a duplicate is found, we must remove already processed argument symbols
+                if (duplicatePos > 0) {
+                    for (Map.Entry<AnalyzedRelation, List<Symbol>> entry : symbolsPerRelation.entrySet()) {
+                        List<Symbol> symbols = entry.getValue();
+                        if (symbols.size() > duplicatePos) {
+                            symbols.remove(duplicatePos);
+                        }
+                    }
+                }
+                return null;
+            }
+            return super.visitFunction(function, symbolsPerRelation);
+        }
+    }
+
+    public static class RelationDetector extends SymbolVisitor<Void, AnalyzedRelation> {
+
+        @Override
+        public AnalyzedRelation visitFunction(Function symbol, Void context) {
+            return symbol.arguments().stream().findFirst().map(arg -> process(arg, context)).orElse(null);
```

Comment:
why is it findFirst() here? Will it work if you have `1 + t1.a` for example?
(maybe I'm just confused)

Reply:
right, will fix